### PR TITLE
feat: integrate DID pallet in ctype, delegation, and attestation pallets

### PR DESCRIPTION
**Pull request by Diiaablo95**
## DID pallet integration with all the other Kilt pallets

Fixes:

- KILTProtocol/ticket#1256
- KILTProtocol/ticket#1251
- KILTProtocol/ticket#1245
- KILTProtocol/ticket#1205
- KILTProtocol/ticket#1206
- KILTProtocol/ticket#1207
- KILTProtocol/ticket#1142

The main change is the integration of the new DID pallet for the ctype, delegation, and attestation pallets. All mention of `Account` have been replaced with `DID`.

Other changes include:

- update ctype, delegation and attestation pallets to use the latest FRAME version for both pallet and test code
- improved unit test coverage for ctype, delegation, and attestations pallets
- fix some bugs in the delegation pallet, especially regarding the verification of the `is_delegating` condition (see tickets above)
- fix a small typo for the DID key agreement keys, which was previously X55519, and now it has been renamed to X25519 (thanks @rflechtner)
- 

## How to test:

Run `cargo test --all --all-features`

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

**Comment by weichweich** [comment]
TLDR! 10K lines?! 😭

**Comment by Diiaablo95** [comment]
@weichweich a good 40% of them are very detailed test cases 😂

**Comment by Diiaablo95** [comment]
While talking with Raphael, I notice a couple of things still missing in the DID pallet, specifically, the possibility to remove delegation and attestation keys in a DID update operation. I think it would be best to integrate those into this PR as well, but if you think I should open another one for that, we can discuss about it. So if we go the former path, I would say to wait a bit longer to review this PR.

File: pallets/did/src/lib.rs [hunk 1]
```diff
@@ -671,187 +765,208 @@ pub mod pallet {
...
+		let mut did_details =
+			<Did<T>>::get(&operation.get_did()).ok_or(DidError::StorageError(StorageError::DidNotPresent))?;
+
+		Self::verify_operation_validity_for_did(operation, &signature, &did_details)?;
 
-		Self::verify_operation_signature_for_entry(op, signature, &did_entry)?;
+		// Update tx counter in DID details and save to DID pallet
+		did_details
+			.increase_tx_counter()
+			.expect("Increasing DID tx counter should be a safe operation.");
```
Comment (weichweich) [inline pallets/did/src/lib.rs:903]: if it's infallible it shouldn't return a result. If it's fallible we should return the error.

File: pallets/did/src/lib.rs [hunk 2]
```diff
@@ -671,187 +765,208 @@ pub mod pallet {
...
+		ensure!(
+			did_details.get_tx_counter_value() < u64::MAX,
+			DidError::StorageError(StorageError::MaxTxCounterValue)
+		);
+
+		// Verify that the operation counter is equal to the stored one + 1
+		let expected_nonce_value = did_details
+			.get_tx_counter_value()
+			.checked_add(1)
+			.expect("Checked against overflow in previous step.");
```
Comment (weichweich) [inline pallets/did/src/lib.rs:944]: maybe don't panic but return an error.

**Review by weichweich — CHANGES_REQUESTED** [review]
not yet through with the review. But i stop here, since there might be more work in the pipeline. 😁

File: pallets/kilt-launch/src/mock.rs [hunk 3]
```diff
@@ -16,6 +16,8 @@
 
 // If you feel like getting in touch with us, you can do so at info@botlabs.org
 
+#![allow(clippy::from_over_into)]
```
Comment (Diiaablo95) [inline pallets/kilt-launch/src/mock.rs:19]: When running clippy I found a couple of warning, so I addressed them as well.

**Comment by Diiaablo95** [comment]
@weichweich @wischli PR finally ready for review! ✌🏻✌🏻✌🏻

File: pallets/delegation/src/lib.rs [hunk 4]
```diff
@@ -22,490 +22,506 @@
...
+		// Add all values to an u8 vector
 		let mut hashed_values: Vec<u8> = delegation_id.as_ref().to_vec();
 		hashed_values.extend_from_slice(root_id.as_ref());
 		if let Some(parent) = parent_id {
 			hashed_values.extend_from_slice(parent.as_ref())
 		}
 		hashed_values.extend_from_slice(permissions.as_u8().as_ref());
-		// hash vector
+		// Hash the resulting vector
 		T::Hashing::hash(&hashed_values)
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/delegation/src/lib.rs:411]: Let's do that sometime. But not now. 🙂

File: pallets/did/src/types.rs [hunk 5]
```diff
@@ -0,0 +1,990 @@
+// KILT Blockchain – https://botlabs.org
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/types.rs:1]: Shouldn't we then do the same for all the other pallets as well?

File: pallets/did/src/lib.rs [hunk 6]
```diff
@@ -670,188 +178,257 @@ pub mod pallet {
...
+		///   execution fees
+		#[pallet::weight(0)]
+		pub fn submit_did_call(
+			origin: OriginFor<T>,
+			did_call: Box<DidAuthorizedCallOperation<T>>,
+			signature: DidSignature,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			let did_identifier = did_call.did.clone();
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/lib.rs:312]: Yes you are right. I went a bit too far with removing the clones. 😁 So let's keep the Event content. This also only adds one `clone` back. 👍🏻

File: pallets/did/src/types.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/types.rs:1]: Mh eg. [ctypetypes](https://github.com/KILTprotocol/mashnet-node/blob/aa-did-integration/pallets/ctype/src/types.rs) could still fit into `lib.rs` IMHO.

[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/types.rs:1]: yes renaming the delegation/types.rs would also be nice. I would move the type alias to lib.rs and call it delegation_tree.rs

File: pallets/did/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/lib.rs:312]: `let DidAuthorizedCallOperation { did, call, .. } = wrapped_operation.operation;` generates a new clippy warning about the use of `Box`, which is definitely not correct, as if not used it creates a memory cycle. I guess we can keep the original solution and use `clone()` two times?

[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/lib.rs:312]: `local variable doesn't need to be boxed here`

**Review by weichweich — APPROVED** [review]
Very nice! I really like the new DID pallet! 🙂 Great work!

**Review by wischli — APPROVED** [review]
Also looks good to me. Great stuff!
